File: HandSock-Server/src/main/java/com/server/handsock/sockets/handler/UserAuthHandler.java
package com.server.handsock.sockets.handler;

import com.corundumstudio.socketio.AckRequest;
import com.corundumstudio.socketio.SocketIOClient;
import com.corundumstudio.socketio.SocketIOServer;
import com.server.handsock.admin.service.ServerSystemService;
import com.server.handsock.console.ConsolePrints;
import com.server.handsock.console.HandleResults;
import com.server.handsock.console.UtilityService;
import com.server.handsock.services.ClientService;
import com.server.handsock.services.TokenService;
import com.server.handsock.clients.service.ClientUserService;
import com.server.handsock.sockets.eventer.OnlineEvent;

import java.util.Map;

public class UserAuthHandler {

    private final Map<String, Object> service;
    private final Map<String, Object> serverServiceList;
    private final Map<String, Object> clientServiceList;
    private final UtilityService utilityService = new UtilityService();
    private final HandleResults handleResults = new HandleResults();

    public UserAuthHandler(Map<String, Object> service, Map<String, Object> clientServiceList, Map<String, Object> serverServiceList) {
        this.service = service;
        this.clientServiceList = clientServiceList;
        this.serverServiceList = serverServiceList;
    }

    public void handleUserLogin(SocketIOClient client, Map<String, Object> data, OnlineEvent onlineEvent, AckRequest ackSender) {
        ClientService clientService = new ClientService(service, clientServiceList);
        try {
            if (!onlineEvent.clientList.contains(utilityService.encodeStringToMD5(client.getSessionId().toString()))) {
                ackSender.sendAckData(handleResults.handleResultByCode(403, null, "禁止访问"));
            } else {
                ClientUserService clientUserService = (ClientUserService) clientServiceList.get("clientUserService");
                ackSender.sendAckData(clientUserService.loginUser(clientService.getClientData(data, "username"), clientService.getClientData(data, "password"), clientService.getRemoteAddress(client)));
            }
        } catch (Exception e) {
            clientService.handleException(e, ackSender);
        }
    }

    public void handleUserLogout(SocketIOClient client, SocketIOServer server, OnlineEvent onlineEvent, AckRequest ackSender) {
        ClientService clientService = new ClientService(service, clientServiceList);
        if (clientService.validClientToken(client)) {
            try {
                if (!onlineEvent.clientList.contains(utilityService.encodeStringToMD5(client.getSessionId().toString()))) {
                    ackSender.sendAckData(handleResults.handleResultByCode(403, null, "禁止访问"));
                } else {
                    TokenService tokenService = (TokenService) service.get("tokenService");
                    tokenService.removeUserToken(clientService.getRemoteUID(client));
                    onlineEvent.sendUserDisconnect(server, client, clientService.getRemoteAddress(client));
                    new ConsolePrints().printInfoLogV2("User Logout " + clientService.getRemoteAddress(client) + " " + clientService.getRemoteUID(client));
                }
            } catch (Exception e) {
                clientService.handleException(e, ackSender);
            }
        } else ackSender.sendAckData(handleResults.handleResultByCode(403, null, "已是退出状态"));
    }

    public void handleUserRegister(SocketIOServer server, SocketIOClient client, Map<String, Object> data, OnlineEvent onlineEvent, AckRequest ackSender) {
        ClientService clientService = new ClientService(service, clientServiceList);
        try {
            if (!onlineEvent.clientList.contains(utilityService.encodeStringToMD5(client.getSessionId().toString()))) {
                ackSender.sendAckData(handleResults.handleResultByCode(403, null, "禁止访问"));
            } else {
                ServerSystemService serverSystemService = (ServerSystemService) serverServiceList.get("serverSystemService");
                if (serverSystemService.getSystemKeyStatus("register")) {
                    ClientUserService clientUserService = (ClientUserService) clientServiceList.get("clientUserService");
                    Map<String, Object> result = clientUserService.registerUser(clientService.getClientData(data, "username"), clientService.getClientData(data, "password"), clientService.getRemoteAddress(client));
                    ackSender.sendAckData(result);
                    if (Integer.parseInt(clientService.getClientData(result, "code")) == 200) utilityService.sendGlobalMessage(server, "[RE:USER:ALL]", null);
                } else ackSender.sendAckData(handleResults.handleResultByCode(402, null, "当前禁止注册"));
            }
        } catch (Exception e) {
            clientService.handleException(e, ackSender);
        }
    }
}

File: HandSock-Server/src/main/java/com/server/handsock/sockets/handler/UserAuthHandler.kt
package com.server.handsock.sockets.handler

import com.corundumstudio.socketio.AckRequest
import com.corundumstudio.socketio.SocketIOClient
import com.corundumstudio.socketio.SocketIOServer
import com.server.handsock.admin.service.ServerSystemService
import com.server.handsock.clients.service.ClientUserService
import com.server.handsock.services.AuthService
import com.server.handsock.services.ClientService
import com.server.handsock.services.TokenService
import com.server.handsock.sockets.eventer.OnlineEvent
import com.server.handsock.utils.ConsoleUtils
import com.server.handsock.utils.HandUtils
import com.server.handsock.utils.IDGenerator
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class UserAuthHandler @Autowired constructor(
    private val authService: AuthService,
    private val onlineEvent: OnlineEvent,
    private val tokenService: TokenService,
    private val clientService: ClientService,
    private val clientUserService: ClientUserService,
    private val serverSystemService: ServerSystemService
) {
    fun handleUserLogin(client: SocketIOClient, data: Map<String?, Any>, ackSender: AckRequest) {
        try {
            if (!onlineEvent.checkClient(HandUtils.encodeStringToMD5(client.sessionId.toString()))) {
                ackSender.sendAckData(HandUtils.handleResultByCode(403, null, "禁止访问"))
            } else {
                ackSender.sendAckData(
                    clientUserService.loginUser(
                        address = clientService.getRemoteAddress(client),
                        username = clientService.getClientData(data, "username"),
                        password = clientService.getClientData(data, "password")
                    )
                )
            }
        } catch (e: Exception) {
            ackSender.sendAckData(HandUtils.printErrorLog(e))
        }
    }

    fun handleUserLogout(client: SocketIOClient, server: SocketIOServer?, ackSender: AckRequest) {
        ackSender.sendAckData(authService.validClientStatusBySocket(client) {
            if (!onlineEvent.checkClient(HandUtils.encodeStringToMD5(client.sessionId.toString()))) {
                HandUtils.handleResultByCode(403, null, "禁止访问")
            } else {
                onlineEvent.sendUserDisconnect(server!!, client)
                ConsoleUtils.printInfoLog(
                    "User Logout ${clientService.getRemoteAddress(client)} ${clientService.getRemoteUID(client)}"
                )
            }
        })
    }

    fun handleUserRegister(client: SocketIOClient, data: Map<String?, Any>, ackSender: AckRequest) {
        try {
            if (!onlineEvent.checkClient(HandUtils.encodeStringToMD5(client.sessionId.toString()))) {
                ackSender.sendAckData(HandUtils.handleResultByCode(403, null, "禁止访问"))
            } else {
                if (serverSystemService.getSystemKeyStatus("register")) {
                    val result = clientUserService.registerUser(
                        clientService.getClientData(data, "username"),
                        clientService.getClientData(data, "password"),
                        clientService.getRemoteAddress(client)
                    )
                    ackSender.sendAckData(result)
                } else ackSender.sendAckData(HandUtils.handleResultByCode(402, null, "当前禁止注册"))
            }
        } catch (e: Exception) {
            ackSender.sendAckData(HandUtils.printErrorLog(e))
        }
    }

    fun handleUserScanLogin(client: SocketIOClient, ackSender: AckRequest) {
        try {
            if (!onlineEvent.checkClient(HandUtils.encodeStringToMD5(client.sessionId.toString()))) {
                ackSender.sendAckData(HandUtils.handleResultByCode(403, null, "禁止访问"))
            } else {
                val qid = IDGenerator.generateUniqueId()
                val content = mapOf(
                    "qid" to qid,
                    "type" to "login",
                    "platform" to "H5"
                )
                tokenService.setScanStatus(qid, 0)
                /*QrcodeUtils.generateQrcode(content.toString())*/
                ackSender.sendAckData(HandUtils.handleResultByCode(200, content, "获取成功"))
            }
        } catch (e: Exception) {
            ackSender.sendAckData(HandUtils.printErrorLog(e))
        }
    }

    fun handleGetScanLoginStatus(client: SocketIOClient, data: Map<String?, Any>, ackSender: AckRequest) {
        try {
            if (!onlineEvent.checkClient(HandUtils.encodeStringToMD5(client.sessionId.toString()))) {
                ackSender.sendAckData(HandUtils.handleResultByCode(403, null, "禁止访问"))
            } else {
                val qid = clientService.getClientData(data, "qid")
                ackSender.sendAckData(clientUserService.getUserQrcodeScanStatus(
                    qid = qid,
                    address = clientService.getRemoteAddress(client)
                ))
            }
        } catch (e: Exception) {
            ackSender.sendAckData(HandUtils.printErrorLog(e))
        }
    }
}